### PR TITLE
Make AI hero task management

### PR DESCRIPTION
**Pull request by ihhub**
- don't hire heroes if no tasks exist
- deprioterize monster strength if no tasks exist
- re-evaluate Barbarian castle development
- fix incorrect neareast tile to move logic

### CODE REVIEW

File: src/fheroes2/world/world_pathfinding.cpp [hunk 1]
```diff
@@ -453,7 +453,7 @@ int AIWorldPathfinder::getNeareastTileToMove( const Heroes & hero )
                 continue;
 
             const MapsIndexes & monsters = Maps::GetTilesUnderProtection( newIndex );
-            if ( _cache[newIndex]._cost && monsters.empty() ) {
+            if ( monsters.empty() && world.getPath( hero, newIndex ).size() == 1 && world.GetTiles( newIndex ).isClearGround() ) {
```
Comment (oleg-derevenetz) [inline src/fheroes2/world/world_pathfinding.cpp:456]: Why this change? Isn't checking `_cache[newIndex]._cost` enough? While calculating cost, passability rules should be taken into account, isn't it?

Comment (ihhub) [reply src/fheroes2/world/world_pathfinding.cpp:456]: Hmmm... I encountered some issues not detecting a neighbouring tile as valid. I've also found a huge issue with this funnction as well.. Give me few minutes to fix it.

Comment (ihhub) [reply src/fheroes2/world/world_pathfinding.cpp:456]: Found the root of the problem: we shuffle directions but use `_mapOffset` for unshuffled case: line 451.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub it's up to you (as always 😀), but I'd be postponed these changes after 0.9.5 to properly test them.

**Comment by ihhub** [comment]
> Hi @ihhub it's up to you (as always ), but I'd be postponed these changes after 0.9.5 to properly test them.

I actually found issues in AI and that's why I pushed these changes. I've tested like 3 hours for different maps. So far so good.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I had to rush a bit with today's pull requests as I noticed that our progress was a bit slow :( I hope for the next release we won't have such situation.

**Comment by ihhub** [comment]
By our progress I mean my progress as I delayed multiple pull request reviews since I was busy in real life...

**Comment by oleg-derevenetz** [comment]
Hi @ihhub are we in a hurry somewhere? :)

**Comment by ihhub** [comment]
I compared the progress in relation with other releases. Of course we aren't in hurry anywhere but we still better stick to our schedule :) Anyway, I can see that the amount of issues is reducing over time which means we're on the right track.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub Well, it's a volunteer project after all, so progress will be always determined by the presence of a free time :) It's inevitable.
